File: src/BlockType.cpp
#include <BlockType.hpp>

BlockType::BlockType(TileType* _tile_type, const std::vector<std::vector<bool>> _grid, bool _rotate) {
	tile_type = _tile_type;
	grid = _grid;
	rotate = _rotate;
	// Used for alignment in "next block" area
	width = 0;
	starting_line = 0;
	for (unsigned int y = 0; y < grid.size(); y++) {
		bool has_content = false;
		for (unsigned int x = 0; x < grid[y].size(); x++) {
			if (grid[y][x]) {
				width = std::max({width, x + 1});
				has_content = true;
			}
		}
		if (has_content) {
			if (height == 0) {
				starting_line = y;
			}
			height = y + 1 - starting_line;
		}
	}
}

BlockType* BlockType::random() {
	init();
	return &list[rand() % list.size()];
}

void BlockType::init() {
	if (inited) {
		return;
	}

	// I block
	list.push_back(BlockType(new TileType(
			sf::IntRect(0,             0, TILE_SIZE, TILE_SIZE),
			sf::IntRect(0,             TILE_SIZE, TILE_SIZE, TILE_SIZE)
		), {
			{0, 0, 0, 0},
			{1, 1, 1, 1},
			{0, 0, 0, 0},
			{0, 0, 0, 0}
		}
	));

	// J Block
	list.push_back(BlockType(new TileType(
			sf::IntRect(TILE_SIZE,     0, TILE_SIZE, TILE_SIZE),
			sf::IntRect(TILE_SIZE,     TILE_SIZE, TILE_SIZE, TILE_SIZE)
		), {
			{1, 0, 0},
			{1, 1, 1},
			{0, 0, 0}
		}
	));

	// L Block
	list.push_back(BlockType(new TileType(
			sf::IntRect(TILE_SIZE * 2, 0, TILE_SIZE, TILE_SIZE),
			sf::IntRect(TILE_SIZE * 2, TILE_SIZE, TILE_SIZE, TILE_SIZE)
		), {
			{0, 0, 1},
			{1, 1, 1},
			{0, 0, 0}
		}
	));

	// O Block
	list.push_back(BlockType(new TileType(
			sf::IntRect(TILE_SIZE * 3, 0, TILE_SIZE, TILE_SIZE),
			sf::IntRect(TILE_SIZE * 3, TILE_SIZE, TILE_SIZE, TILE_SIZE)
		), {
			{1, 1},
			{1, 1}
		}, false
	));

	// S Block
	list.push_back(BlockType(new TileType(
			sf::IntRect(TILE_SIZE * 4, 0, TILE_SIZE, TILE_SIZE),
			sf::IntRect(TILE_SIZE * 4, TILE_SIZE, TILE_SIZE, TILE_SIZE)
		), {
			{0, 1, 1},
			{1, 1, 0},
			{0, 0, 0}
		}
	));

	// T Block
	list.push_back(BlockType(new TileType(
			sf::IntRect(TILE_SIZE * 5, 0, TILE_SIZE, TILE_SIZE),
			sf::IntRect(TILE_SIZE * 5, TILE_SIZE, TILE_SIZE, TILE_SIZE)
		), {
			{0, 1, 0},
			{1, 1, 1},
			{0, 0, 0}
		}
	));

	// Z Block
	list.push_back(BlockType(new TileType(
			sf::IntRect(TILE_SIZE * 6, 0, TILE_SIZE, TILE_SIZE),
			sf::IntRect(TILE_SIZE * 6, TILE_SIZE, TILE_SIZE, TILE_SIZE)
		), {
			{1, 1, 0},
			{0, 1, 1},
			{0, 0, 0}
		}
	));

	inited = true;
}

File: src/structs/block_type.rs
use crate::structs::TileType;
use crate::config::TILE_SIZE;
use sfml::graphics::IntRect;

pub struct BlockType {
	pub tile_type: TileType,
	pub grid: Vec<Vec<bool>>,
	pub width: u32,
	pub height: u32,
	pub starting_line: u32,
	pub rotate: bool
}

impl BlockType {
	pub fn new(tile_type: TileType, grid: Vec<Vec<bool>>, rotate: bool) -> Self {
		let mut width: u32 = 0;
		let mut height: u32 = 0;
		let mut starting_line: u32 = 0;
		for (y, row) in grid.iter().enumerate() {
			let mut has_content = false;
			for (x, cell) in row.iter().enumerate() {
				if *cell {
					width = std::cmp::max(width, x as u32 + 1);
					has_content = true;
				}
			}
			if has_content {
				if height == 0 {
					starting_line = y as u32;
				}
				height = y as u32 + 1 - starting_line;
			}
		}
		Self {
			tile_type,
			grid,
			width,
			height,
			starting_line,
			rotate
		}
	}

	pub fn init_list() -> Vec<Self> {
		let mut list = Vec::new();
		let tile_size = TILE_SIZE as i32;

		const Y: bool = true;
		const N: bool = false;

		// I block
		list.push(Self::new(
			TileType::new(
				IntRect::new(0, 0, tile_size, tile_size),
				IntRect::new(0, tile_size, tile_size, tile_size)
			),
			vec![
				vec![N, N, N, N],
				vec![Y, Y, Y, Y],
				vec![N, N, N, N],
				vec![N, N, N, N]
			],
			true
		));

		// J Block
		list.push(Self::new(
			TileType::new(
				IntRect::new(tile_size, 0, tile_size, tile_size),
				IntRect::new(tile_size, tile_size, tile_size, tile_size),
			),
			vec![
				vec![Y, N, N],
				vec![Y, Y, Y],
				vec![N, N, N]
			],
			true
		));

		// L Block
		list.push(Self::new(
			TileType::new(
				IntRect::new(tile_size * 2, 0, tile_size, tile_size),
				IntRect::new(tile_size * 2, tile_size, tile_size, tile_size),
			),
			vec![
				vec![N, N, Y],
				vec![Y, Y, Y],
				vec![N, N, N]
			],
			true
		));

		// O Block
		list.push(Self::new(
			TileType::new(
				IntRect::new(tile_size * 3, 0, tile_size, tile_size),
				IntRect::new(tile_size * 3, tile_size, tile_size, tile_size),
			),
			vec![
				vec![Y, Y],
				vec![Y, Y]
			],
			false
		));

		// S Block
		list.push(Self::new(
			TileType::new(
				IntRect::new(tile_size * 4, 0, tile_size, tile_size),
				IntRect::new(tile_size * 4, tile_size, tile_size, tile_size),
			),
			vec![
				vec![N, Y, Y],
				vec![Y, Y, N],
				vec![N, N, N]
			],
			true
		));

		// T Block
		list.push(Self::new(
			TileType::new(
				IntRect::new(tile_size * 5, 0, tile_size, tile_size),
				IntRect::new(tile_size * 5, tile_size, tile_size, tile_size),
			),
			vec![
				vec![N, Y, N],
				vec![Y, Y, Y],
				vec![N, N, N]
			],
			true
		));

		// Z Block
		list.push(Self::new(
			TileType::new(
				IntRect::new(tile_size * 6, 0, tile_size, tile_size),
				IntRect::new(tile_size * 6, tile_size, tile_size, tile_size)
			),
			vec![
				vec![Y, Y, N],
				vec![N, Y, Y],
				vec![N, N, N]
			],
			true
		));
		
		list
	}
}